### PR TITLE
update-bash: git related improvement

### PR DESCRIPTION
**Pull request by xu-cheng**
* Use git function instead of refreshing bash cache on `git` path.
* Better `which_git`:
  * Set `HOMEBREW_GIT` as result.
  * Take user's setting of `HOMEBREW_GIT` and `GIT` env variable into
    account.

cc @mikemcquaid @UniqMartin

### CODE REVIEW

File: Library/Homebrew/cmd/update-bash.sh [hunk 1]
```diff
@@ -2,22 +2,38 @@ brew() {
   "$HOMEBREW_BREW_FILE" "$@"
 }
 
+git() {
+  "$HOMEBREW_GIT" "$@"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: Might want to run `which_git` if this isn't already set.

File: Library/Homebrew/cmd/update-bash.sh [hunk 2]
```diff
@@ -2,22 +2,38 @@ brew() {
...
+  fi
+
+  if [[ -n "$GIT" && -x "$GIT" ]]
+  then
+    HOMEBREW_GIT="$GIT"
+    echo "$HOMEBREW_GIT"
+    return
+  fi
+
+  HOMEBREW_GIT="$(which git 2>/dev/null)"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: I think it would be nice to maybe have some of this as a set environment variable in `bin/brew`.

File: Library/Homebrew/cmd/update-bash.sh [hunk 3]
```diff
@@ -271,7 +287,6 @@ EOS
       odie "Git must be installed and in your PATH!"
     fi
   fi
-  hash -p "$(cd "$(dirname "$(which_git)")" && pwd -P)/git" git
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: You're removing this but you're not actually resolving the full path at any point so it'll break when you change directory.

File: Library/Homebrew/cmd/update-bash.sh (hunk 2 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/update-bash.sh]: I think the control flow in this part should be more along the lines of

```shell
if [[ -n "$HOMEBREW_GIT" ]]
then
  : # We'll take the user-provided `HOMEBREW_GIT` directly.
elif [[ -n "$GIT" ]]
then
  HOMEBREW_GIT="$GIT"
else
  HOMEBREW_GIT="$(which git 2>/dev/null)"
fi
```

or condensed to a single line (but maybe deemed too obscure)

```shell
: "${HOMEBREW_GIT:=${GIT:-$(which git 2>/dev/null)}}"
```

followed by the special handling of `/usr/bin/git` and the active developer directory (we should be doing this no matter where the `git` path came from), finally followed by a check if the determined path exists and is executable.

Comment (UniqMartin) [inline Library/Homebrew/cmd/update-bash.sh]: Given the task this function is now performing I think it should be transformed into the shell equivalent of `Utils.ensure_git_installed!` and invoked at the beginning of a shell command that requires `git`. And the newly introduced `git` shell function itself might be best off failing hard if this setup function was never executed.

> I think it would be nice to maybe have some of this as a set environment variable in `bin/brew`.

:+1: If we can do that in a way that doesn't slow down commands that don't require `git` at all.

Comment (xu-cheng) [inline Library/Homebrew/cmd/update-bash.sh]: > > I think it would be nice to maybe have some of this as a set environment variable in bin/brew.

> :+1: If we can do that in a way that doesn't slow down commands that don't require git at all.

I am not sure this is a good idea because not all of commands require `git` at all. Even more, for commands which require `git`, the ruby version of `Utils.ensure_git_installed!` and `Utils.git_available?` are still required. Setting the environment can be confusing in such circumstance.

How about creating a `Library/Utils/git.sh` file which is a bash equivalent of `Library/Homebrew/utils/git.rb` and only source it when needed for bash commands.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: > How about creating a Library/Utils/git.sh file which is a bash equivalent of Library/Homebrew/utils/git.rb and only source it when needed by bash commands.

If that's the case: I'd rather leave it as-is for now rather than splitting it up before it's necessary to do so. Let's wait until we have more Bash commands :grinning:

**Comment by xu-cheng** [comment]
Updated and tested locally.

File: Library/Homebrew/cmd/update-bash.sh [hunk 4]
```diff
@@ -2,22 +2,41 @@ brew() {
   "$HOMEBREW_BREW_FILE" "$@"
 }
 
+git() {
+  [[ -z "$HOMEBREW_GIT" ]] && HOMEBREW_GIT="$(which_git 2>/dev/null)"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: I think this should just error out if it's empty; it won't produce an error as-is.

File: Library/Homebrew/cmd/update-bash.sh [hunk 5]
```diff
@@ -263,15 +282,16 @@ EOS
     odie "$HOMEBREW_REPOSITORY must be writable!"
   fi
 
-  if [[ -z "$(which_git)" ]]
+  HOMEBREW_GIT="$(which_git)"
+  if [[ -z "$HOMEBREW_GIT" ]]
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: Can you do the assignment inside the `[[]]` somehow? I didn't think so but figured I'd ask CC @UniqMartin

Comment (UniqMartin) [inline Library/Homebrew/cmd/update-bash.sh]: If that's possible, I don't know how.

File: Library/Homebrew/cmd/update-bash.sh (hunk 4 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/update-bash.sh]: I wonder that we should add `set -e` in `bin/brew`. It's important especially in subshell.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: The problem is it's not very user friendly; things will just fail and it's nearly impossible to tell what and why.

Comment (xu-cheng) [inline Library/Homebrew/cmd/update-bash.sh]: But without it,  error here when git  is run in subshell will have strange result.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: Yeh, I get that, I just think it's a pretty major change that should be handled in another PR along with e.g. a trap to improve the messaging.

Comment (xu-cheng) [inline Library/Homebrew/cmd/update-bash.sh]: :cool:

**Comment by xu-cheng** [comment]
Any comment before I :ship:

File: Library/Homebrew/cmd/update-bash.sh [hunk 6]
```diff
@@ -2,22 +2,41 @@ brew() {
...
+  elif [[ -n "$GIT" && -x "$GIT" ]]
+  then
+    git_path="$GIT"
+  else
+    git_path="$(which git 2>/dev/null)"
+  fi
+
+  if [[ -n "$git_path" ]]
+  then
+    git_path="$(chdir "${git_path%/*}" && pwd -P)/${git_path##*/}"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-bash.sh]: Are you sure `${git_path%/*}` behaves the same as `dirname`? I'd probably rather we used `dirname` and the raw `git` string here; the performance implications are minimal (this is not a critical path) and the readability is better using them.

**Comment by mikemcquaid** [comment]
One comment otherwise :+1:

File: Library/Homebrew/cmd/update-bash.sh (hunk 6 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/update-bash.sh]: > Are you sure ${git_path%/*} behaves the same as dirname?

In fact, we use this for years in `bin/brew`.

Comment (UniqMartin) [inline Library/Homebrew/cmd/update-bash.sh]: It behaves differently if the string doesn't contain a slash at all, but otherwise it's doing the same. For the mentioned case `dirname` will return `.` while `${git_path%/*}` will return the original string.

Comment (xu-cheng) [inline Library/Homebrew/cmd/update-bash.sh]: OK, I will change it to `dirname` when pulling.

File: Library/Homebrew/cmd/update-bash.sh [hunk 7]
```diff
@@ -2,22 +2,41 @@ brew() {
   "$HOMEBREW_BREW_FILE" "$@"
 }
 
+git() {
+  [[ -n "$HOMEBREW_GIT" ]] || exit 1
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/update-bash.sh]: Why not use `odie` instead?

File: Library/Homebrew/cmd/update-bash.sh (hunk 6 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/update-bash.sh]: > OK, I will change it to `dirname` when pulling.

That's not what I was trying to imply. Does it make sense to have something like `git-2.2` in one of the variables, with the expectation that it will be looked up in `PATH`? If yes, the whole construct doesn't make sense and will fail, no matter if we use `dirname` or not.

Comment (xu-cheng) [inline Library/Homebrew/cmd/update-bash.sh]: How about:
```
  if [[ -n "$HOMEBREW_GIT" ]]
  then
    git_path="$HOMEBREW_GIT"
  elif [[ -n "$GIT" ]]
  then
    git_path="$GIT"
  else
    git_path="git"
  fi

  git_path="$(which "$git_path" 2>/dev/null)"
```

Comment (UniqMartin) [inline Library/Homebrew/cmd/update-bash.sh]: This looks like something that would work in all scenarios I can currently think of. And it should be guaranteed to return a path that contains a slash. :+1:
